src/mem/mem_ctrl.cc

```diff
@@ -302,10 +460,90 @@ MemCtrl::addToReadQueue(PacketPtr pkt, unsigned int pkt_count, bool is_dram)
     // queue, do so now
     if (!nextReqEvent.scheduled()) {
         DPRINTF(MemCtrl, "Request scheduled immediately\n");
+        //COMMENT: DRAM cache tag check latency should be added to the
+        // curTick()
+        schedule(nextReqEvent, curTick());
+        //COMMENT: tags will be checked on the way back
+    }
+}
+
+void
+MemCtrl::addToDRAMFillQueue(MemPacket mem_pkt)
+{
+    // this is the packet that came from resp queue
+    // and is sent ot nvm read queue (if it did not come from
+    // nvm read queue already)
+
+    //MARYAM: I guess the next line must assert for WRITES, not READS!
+    //assert(mem_pkt->isWrite());
+    assert(!dramFillQueueFull(1));
+
+    // COMMENT: Should overwrite the mem_pkt?
+    // COMMENT: Currently, we are assuming that
+    // a pkt will be decomposed in only one mem_pkt
+    // which probably is not a reasonable assumption
+    mem_pkt = dram->decodePacket(mem_pkt->pkt, mem_pkt->pkt->getAddr(),
+                            mem_pkt->pkt->getSize(), false, false);
+    dram->setupRank(mem_pkt->rank, false);
+    mem_pkt->readyTime = MaxTick;
+
+    // the mem_pkt needs to become a write request
+    // now
+    dramFillQueue.push_back(mem_pkt);
+
+    dramFillQueueSize++;
+
+    // update the DRAM tags as well
+
+    int index = bits(mem_pkt->pkt->getAddr(),
+            ceilLog2(64)+ceilLog2(numEntries), ceilLog2(64));
+
+    tagStoreDC[index].tag = returnTag(mem_pkt->pkt->getAddr());
+
+    // make sure that the block is set to be valid and clean
+    // MARYAM: only valid bit is set here. How about dirty bit?
+    tagStoreDC[index].metadata = 0x01;
+
+    //isInWriteQueue.insert(burstAlign(addr, is_dram));
+
+    if (!nextReqEvent.scheduled()) {
+        // We might need to add tag check latency
+        // because the NVM response might check tags
+        // to see if dram miss was clean or dirty
         schedule(nextReqEvent, curTick());
     }
 }
 
+void
+MemCtrl::addToNVMWriteQueue(MemPacket* mem_pkt)
+{
+    assert(pkt->isWrite());
+    assert(!isNVMWriteQueueFull());
+
+    // COMMENT: Should overwrite the mem_pkt?
+    // COMMENT: Currently, we are assuming that
+    // a pkt will be decomposed in only one mem_pkt
+    // which probably is not a reasonable assumption
+    mem_pkt = nvm->decodePacket(mem_pkt->pkt, mem_pkt->pkt->getAddr(),
+                            mem_pkt->pkt->getSize(), true, false);
+    nvm->setupRank(mem_pkt->rank, true);
+    mem_pkt->readyTime = MaxTick;
+
+    nvmWriteQueue.push_back(mem_pkt);
+
+    nvmWriteQueueSize++;
+
+    if (!nextReqEvent.scheduled()) {
+    // COMEMNT: scheduling a request if it has not been
+    // previously scheduled.
+    // If something else has scheduled nextReqEvent, how would
+    // we add tagCheckLatency. And does it even matter to add that
+    schedule(nextReqEvent, curTick() + tagCheckLatency);
```

Comment:
whitespace
```suggestion
        schedule(nextReqEvent, curTick() + tagCheckLatency);
```

Either add a parameter to this function with latency or explain why you add tagCheckLatency here and not in the other fill/add to queue functions.